File: src/lib/api/customer360.ts
import { apiGet } from '@/lib/api/client.ts'

export type Customer360Partner = {
  id: string
  name: string
  short_name: string
  category: string
  tag: string
  tax_registration_no: string
  bank_name: string
  bank_account: string
  invoice_content: string
  recipient: string
  recipient_phone: string
  mailing_address: string
  created_at: string | null
  updated_at: string | null
}

export type Customer360Access = {
  contracts: boolean
  reconciliation: boolean
  invoices: boolean
  funds: boolean
}

export type Customer360Summary = {
  contract_count: number | null
  active_contract_count: number | null
  contract_amount: number | null
  rd_bill_count: number | null
  rd_settlement_amount: number | null
  rd_paid_amount: number | null
  rd_unpaid_amount: number | null
  channel_bill_count: number | null
  channel_settlement_amount: number | null
  channel_received_amount: number | null
  channel_unreceived_amount: number | null
  invoice_count: number | null
  invoice_amount: number | null
  input_invoice_count: number | null
  output_invoice_count: number | null
  bank_transaction_count: number | null
  bank_inflow_amount: number | null
  bank_outflow_amount: number | null
  latest_trade_date: string | null
}

export type Customer360Contract = {
  id: string
  internal_contract_no: string
  contract_no: string
  contract_name: string
  contract_type: string
  products: string[]
  channels: string[]
  amount: number
  effective_date: string | null
  end_date: string | null
  performance_status: string
  payment_type: string
  state: 'active' | 'pending' | 'expired' | 'ended'
  created_at: string | null
  updated_at: string | null
}

export type Customer360RdBill = {
  id: string
  statement_no: string
  settlement_month: string
  games: string
  settlement_amount: number
  paid_amount: number
  unpaid_amount: number
  payment_status: string
  latest_payment_date: string | null
  status: string
  created_at: string | null
  updated_at: string | null
}

export type Customer360ChannelBill = {
  id: string
  statement_no: string
  settlement_month: string
  games: string
  settlement_amount: number
  received_amount: number
  unreceived_amount: number
  receipt_status: string
  status: string
  created_at: string | null
  updated_at: string | null
}

export type Customer360Invoice = {
  id: string
  direction: 'input' | 'output'
  invoice_no: string
  invoice_date: string
  buyer_name: string
  seller_name: string
  amount: number
  tax_amount: number
  tax_status: string
  status: string
  created_at: string | null
  updated_at: string | null
}

export type Customer360BankTransaction = {
  id: string
  type: string
  trade_date: string
  transaction_no: string
  payer_name: string
  payee_name: string
  summary: string
  inflow: number
  outflow: number
  currency: string
  reconciliation_no: string
  status: string
  created_at: string | null
  updated_at: string | null
}

export type Customer360Activity = {
  kind: 'contract' | 'rd_bill' | 'channel_bill' | 'invoice' | 'bank_transaction'
  entity_id: string
  date: string | null
  title: string
  amount: number
  meta: string
}

export type Customer360Response = {
  partner: Customer360Partner
  access: Customer360Access
  summary: Customer360Summary
  contracts: Customer360Contract[]
  rd_bills: Customer360RdBill[]
  channel_bills: Customer360ChannelBill[]
  invoices: Customer360Invoice[]
  bank_transactions: Customer360BankTransaction[]
  recent_activities: Customer360Activity[]
}

export function getCustomer360(partnerId: string) {
  return apiGet<Customer360Response>(
    `/api/workbench/customer-360/${encodeURIComponent(partnerId)}`
  )
}

File: backend/app/schemas/customer360.py
"""Response models for V2.5-3 customer 360."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class Customer360Partner(BaseModel):
    id: str
    name: str
    short_name: str = ""
    category: str = ""
    tag: str = ""
    tax_registration_no: str = ""
    bank_name: str = ""
    bank_account: str = ""
    invoice_content: str = ""
    recipient: str = ""
    recipient_phone: str = ""
    mailing_address: str = ""
    created_at: str | None = None
    updated_at: str | None = None


class Customer360Access(BaseModel):
    contracts: bool = False
    reconciliation: bool = False
    invoices: bool = False
    funds: bool = False


class Customer360Summary(BaseModel):
    contract_count: int | None = None
    active_contract_count: int | None = None
    contract_amount: float | None = None
    rd_bill_count: int | None = None
    rd_settlement_amount: float | None = None
    rd_paid_amount: float | None = None
    rd_unpaid_amount: float | None = None
    channel_bill_count: int | None = None
    channel_settlement_amount: float | None = None
    channel_received_amount: float | None = None
    channel_unreceived_amount: float | None = None
    invoice_count: int | None = None
    invoice_amount: float | None = None
    input_invoice_count: int | None = None
    output_invoice_count: int | None = None
    bank_transaction_count: int | None = None
    bank_inflow_amount: float | None = None
    bank_outflow_amount: float | None = None
    latest_trade_date: str | None = None


class Customer360Contract(BaseModel):
    id: str
    internal_contract_no: str = ""
    contract_no: str = ""
    contract_name: str = ""
    contract_type: str = ""
    products: list[str] = Field(default_factory=list)
    channels: list[str] = Field(default_factory=list)
    amount: float = 0
    effective_date: str | None = None
    end_date: str | None = None
    performance_status: str = ""
    payment_type: str = ""
    state: Literal["active", "pending", "expired", "ended"] = "active"
    created_at: str | None = None
    updated_at: str | None = None


class Customer360RdBill(BaseModel):
    id: str
    statement_no: str = ""
    settlement_month: str = ""
    games: str = ""
    settlement_amount: float = 0
    paid_amount: float = 0
    unpaid_amount: float = 0
    payment_status: str = ""
    latest_payment_date: str | None = None
    status: str = ""
    created_at: str | None = None
    updated_at: str | None = None


class Customer360ChannelBill(BaseModel):
    id: str
    statement_no: str = ""
    settlement_month: str = ""
    games: str = ""
    settlement_amount: float = 0
    received_amount: float = 0
    unreceived_amount: float = 0
    receipt_status: str = ""
    status: str = ""
    created_at: str | None = None
    updated_at: str | None = None


class Customer360Invoice(BaseModel):
    id: str
    direction: Literal["input", "output"]
    invoice_no: str = ""
    invoice_date: str = ""
    buyer_name: str = ""
    seller_name: str = ""
    amount: float = 0
    tax_amount: float = 0
    tax_status: str = ""
    status: str = ""
    created_at: str | None = None
    updated_at: str | None = None


class Customer360BankTransaction(BaseModel):
    id: str
    type: str = ""
    trade_date: str = ""
    transaction_no: str = ""
    payer_name: str = ""
    payee_name: str = ""
    summary: str = ""
    inflow: float = 0
    outflow: float = 0
    currency: str = "CNY"
    reconciliation_no: str = ""
    status: str = ""
    created_at: str | None = None
    updated_at: str | None = None


class Customer360Activity(BaseModel):
    kind: Literal["contract", "rd_bill", "channel_bill", "invoice", "bank_transaction"]
    entity_id: str
    date: str | None = None
    title: str = ""
    amount: float = 0
    meta: str = ""


class Customer360Response(BaseModel):
    partner: Customer360Partner
    access: Customer360Access
    summary: Customer360Summary
    contracts: list[Customer360Contract] = Field(default_factory=list)
    rd_bills: list[Customer360RdBill] = Field(default_factory=list)
    channel_bills: list[Customer360ChannelBill] = Field(default_factory=list)
    invoices: list[Customer360Invoice] = Field(default_factory=list)
    bank_transactions: list[Customer360BankTransaction] = Field(default_factory=list)
    recent_activities: list[Customer360Activity] = Field(default_factory=list)
